crates/storage/provider/src/providers/static_file/manager.rs

```diff
@@ -448,6 +452,161 @@ impl StaticFileProvider {
+            // interruption.
+            let mut highest_block = self.get_highest_static_file_block(segment);
+            if initial_highest_block != highest_block {
+                update_unwind_target(highest_block);
+            }
+
+            // Only applies to transaction-based static files. (Receipts & Transactions)
+            //
+            // Make sure the last transaction matches the last block from its indices, since a heal
+            // from a pruning interruption might have decreased the number of transactions without
+            // being able to update the last block of the static file segment.
+            let highest_tx = self.get_highest_static_file_tx(segment);
+            if let Some(highest_tx) = highest_tx {
+                let mut last_block = highest_block.unwrap_or_default();
+                loop {
+                    if let Some(indices) = provider.block_body_indices(last_block)? {
+                        if indices.last_tx_num() <= highest_tx {
+                            break
+                        }
+                    }
+                    if last_block == 0 {
+                        break
+                    }
+                    last_block -= 1;
+
+                    highest_block = Some(last_block);
+                    update_unwind_target(highest_block);
+                }
+            }
+
+            if let Some(unwind) = match segment {
+                StaticFileSegment::Headers => self.check_invariants::<_, tables::Headers>(
+                    provider,
+                    StageId::Headers,
+                    highest_block,
+                    highest_block,
+                )?,
+                StaticFileSegment::Transactions => self
+                    .check_invariants::<_, tables::Transactions>(
+                        provider,
+                        StageId::Bodies,
+                        highest_tx,
+                        highest_block,
+                    )?,
+                StaticFileSegment::Receipts => self.check_invariants::<_, tables::Receipts>(
+                    provider,
+                    StageId::Execution,
+                    highest_tx,
+                    highest_block,
+                )?,
+            } {
+                update_unwind_target(Some(unwind));
+            }
+        }
+
+        Ok(unwind_target.map(PipelineTarget::Unwind))
+    }
+
+    /// Check invariants for each corresponding table and static file segment. See
+    /// [Self::check_consistency] for more.
+    fn check_invariants<TX: DbTx, T: Table<Key = u64>>(
+        &self,
+        provider: &DatabaseProvider<TX>,
+        stage_id: StageId,
+        static_file_entry: Option<u64>,
+        highest_static_block: Option<BlockNumber>,
+    ) -> ProviderResult<Option<BlockNumber>> {
+        let static_file_entry = static_file_entry.unwrap_or_default();
+        let highest_static_block = highest_static_block.unwrap_or_default();
```

Comment:
is it correct that if we don't even have a genesis block in the segment, we unwrap these to `0`?